File: DPL/03/b.cpp
#include<bits/stdc++.h>
using namespace std;

#define MAX 1400

struct Rectangle {
    int height, pos;
};

int H, W;
int buf[MAX][MAX];
int T[MAX][MAX];

int getLargestRectangle(int size, int buf[]) {
    stack<Rectangle> S;
    int maxv = 0;
    buf[size] = 0;

    for (int i = 0; i <= size; i++) {
        Rectangle rect;
        rect.height = buf[i];
        rect.pos = i;
        if (S.empty()) {
            S.push(rect);
        } else {
            if (S.top().height < rect.height) {
                S.push(rect);
            } else if (S.top().height > rect.height) {
                int target = i;
                while (!S.empty() && S.top().height >= rect.height) {
                    Rectangle pre = S.top(); S.pop();
                    int area = pre.height * (i - pre.pos);
                    maxv = max(maxv, area);
                    target = pre.pos;
                }
                rect.pos = target;
                S.push(rect);
            }
        }
    }
    return maxv;
}

int getLargestRectangle() {
    for (int j = 0; j < W; j++) {
        for (int i = 0; i < H; i++) {
            if (buf[i][j]) {
                T[i][j] = 0;
            } else {
                T[i][j] = (i > 0) ? T[i - 1][j] + 1 : 1;
            }
        }
    }

    int maxv = 0;
    for (int i = 0; i < H; i++) {
        maxv = max(maxv, getLargestRectangle(W, T[i]));
    }

    return maxv;
}

int main() {
    scanf("%d %d", &H, &W);
    for (int i = 0; i < H; i++) {
        for (int j = 0; j < W; j++) {
            scanf("%d", &buf[i][j]);
        }
    }
    cout << getLargestRectangle() << endl;
    
    return 0;
}

File: DPL/03/b.py
MAX = 1400


H = W = 0
buf = [[0 for j in xrange(MAX)] for i in xrange(MAX)]
T = [[0 for j in xrange(MAX)] for i in xrange(MAX)]


class Rectangle:

    def __init__(self, h, p):
        self.height = h
        self.pos = p


def getLargestRectangle(size, buf):
    S = []
    maxv = 0
    buf[size] = 0

    for i in xrange(size + 1):
        rect = Rectangle(buf[i], i)
        if len(S) == 0:
            S.append(rect)
        else:
            if S[len(S) - 1].height < rect.height:
                S.append(rect)
            elif S[len(S) - 1].height > rect.height:
                target = i
                while len(S) and S[len(S) - 1].height >= rect.height:
                    pre = S.pop()
                    area = pre.height * (i - pre.pos)
                    maxv = max(maxv, area)
                    target = pre.pos
                rect.pos = target
                S.append(rect)

    return maxv


def solve():
    for j in xrange(W):
        for i in xrange(H):
            if buf[i][j]:
                T[i][j] = 0
            else:
                T[i][j] = (T[i - 1][j] + 1) if i > 0 else 1

    maxv = 0
    for i in xrange(H):
        maxv = max(maxv, getLargestRectangle(W, T[i]))

    return maxv


def main():
    global H, W
    H, W = map(int, raw_input().split())
    for i in xrange(H):
        t = map(int, raw_input().split())
        for j in xrange(W):
            buf[i][j] = t[j]

    print solve()

if __name__ == '__main__':
    main()
